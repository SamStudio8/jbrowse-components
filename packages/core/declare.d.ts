declare module 'shortid'
declare module 'object.fromentries'
