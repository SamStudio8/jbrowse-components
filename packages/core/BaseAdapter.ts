import { Observer, Observable } from 'rxjs'
import { ObservableCreate } from './util/rxjs'
// @ts-ignore
import { checkAbortSignal } from './util'

export interface Region {
  refName: string
  start: number
  end: number
}
interface BaseOptions {
  signal: AbortSignal
}
/**
 * Base class for adapters to extend. Defines some methods that subclasses must
 * implement.
 */
export default class BaseAdapter {
  // List of all possible capabilities. Don't un-comment them here.
  static capabilities = [
    // 'getFeatures',
    // 'getRefNames',
    // 'getRegions',
    // 'getRefNameAliases',
  ]

  constructor(config: any) {
    if (new.target === BaseAdapter) {
      throw new TypeError(
        'Cannot create BaseAdapter instances directly, use a subclass',
      )
    }
  }

  /**
   * Subclasses should override this method. Method signature here for reference.
   * @returns {Promise<string[]>} Array of reference sequence names used by the
   * source being adapted.
   */
  async getRefNames():Promise<string[]> {
    throw new Error('getRefNames should be overridden by the subclass')
    // Subclass method should look something like this:
    // await this.setup()
    // const { refNames } = this.metadata
    // return refNames
  }

  /**
   * Subclasses should override this method. Method signature here for reference.
   * @param {Region} region
   * @param {BaseOptions} options
   * @returns {Observable[Feature]} Observable of Feature objects in the region
   */
  // eslint-disable-next-line no-unused-vars
  getFeatures({ refName, start, end }: Region, opts: BaseOptions):Observable<any> {
    throw new Error('getFeatures should be overridden by the subclass')
    // Subclass method should look something like this:
    // return ObservableCreate(observer => {
    //   const records = getRecords(assembly, refName, start, end)
    //   records.forEach(record => {
    //     observer.next(this.recordToFeature(record))
    //   })
    //   observer.complete()
    // })
  }

  /**
   * Subclasses should override this method. Method signature here for reference.
   *
   * called to provide a hint that data tied to a certain region
   * will not be needed for the forseeable future and can be purged
   * from caches, etc
   * @param {Region} region
   */
  // eslint-disable-next-line no-unused-vars
  freeResources(region: Region) {
    throw new Error('freeResources should be overridden by the subclass')
  }

  /**
   * Checks if the store has data for the given assembly and reference
   * sequence, and then gets the features in the region if it does.
   * @param {Region} region see getFeatures()
   * @param {AbortSignal} [signal] optional AbortSignal for aborting the request
   * @returns {Observable[Feature]} see getFeatures()
   */
  getFeaturesInRegion(region: Region, opts: BaseOptions):Observable<any> {
    return ObservableCreate(async (observer:Observer<any>) => {
      const hasData = await this.hasDataForRefName(region.refName)
      checkAbortSignal(opts.signal)
      if (!hasData) {
        observer.complete()
      } else {
        this.getFeatures(region, opts).subscribe(observer)
      }
    })
  }

  /**
   * Check if the store has data for the given reference name. Also checks the
   * assembly name if one was specified in the initial adapter config.
   * @param {string} refName Name of the reference sequence
   * @returns {Promise<boolean>} Whether data source has data for the given
   * reference name
   */
  async hasDataForRefName(refName:string):Promise<boolean> {
    const refNames = await this.getRefNames()
    if (refNames.includes(refName)) return true
    return false
  }
}
