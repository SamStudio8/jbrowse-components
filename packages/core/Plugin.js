export default class Plugin {
  install() {
    this.installed = true
  }

  configure() {
    this.configured = true
  }
}
