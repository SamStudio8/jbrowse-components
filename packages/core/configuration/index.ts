export {
  ConfigurationSchema,
  ConfigurationSchemaType,
  ConfigurationSchemaDefinition,
  ConfigurationModel,
  ConfigurationReference,
  isConfigurationModel,
  isConfigurationSchemaType,
  isBareConfigurationSchemaType,
} from './configurationSchema'

export { getConf, readConfObject } from './util'
