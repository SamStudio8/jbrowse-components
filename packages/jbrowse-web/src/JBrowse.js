import '@gmod/jbrowse-core/fonts/material-icons.css'
import CircularProgress from '@material-ui/core/CircularProgress'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/styles'
import { observer } from 'mobx-react'
import React from 'react'
import 'typeface-roboto'
import App from './ui/App'
import Theme from './ui/theme'

export default observer(({ state }) => {
  let DisplayComponent = (
    <CircularProgress
      style={{
        position: 'fixed',
        top: '50%',
        left: '50%',
        marginTop: -250,
        marginLeft: -250,
      }}
      size={500}
    />
  )
  if (state.errorMessage) DisplayComponent = <div>{state.errorMessage}</div>
  if (state.activeSession)
    DisplayComponent = (
      <App
        session={state.activeSession}
        sessionNames={state.sessionNames}
        setActiveSession={state.activateSession}
        addSession={state.addSession}
      />
    )

  return (
    <ThemeProvider theme={Theme}>
      <CssBaseline />
      {DisplayComponent}
    </ThemeProvider>
  )
})
