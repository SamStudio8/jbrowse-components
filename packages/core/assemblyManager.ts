import AbortablePromiseCache from 'abortable-promise-cache'
import jsonStableStringify from 'json-stable-stringify'
import { autorun, observable } from 'mobx'
import {
  addDisposer,
  cast,
  getParent,
  IAnyType,
  SnapshotOrInstance,
  types,
} from 'mobx-state-tree'
import { readConfObject } from './configuration'
import { AnyConfigurationModel } from './configuration/configurationSchema'
import { Region } from './util/types'
import { Region as MSTRegion } from './util/types/mst'
import QuickLRU from './util/QuickLRU'

function checkRefName(refName: string) {
  // Valid refName pattern from https://samtools.github.io/hts-specs/SAMv1.pdf
  if (
    !refName.match(
      /[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*/,
    )
  ) {
    throw new Error(`Encountered invalid refName: "${refName}"`)
  }
}

export function assemblyFactory(assemblyConfigType: IAnyType) {
  return types
    .model({
      configuration: types.reference(assemblyConfigType),
      regions: types.array(MSTRegion),
      refNameAliases: types.map(types.array(types.string)),
      refNameAliasesSet: false,
    })
    .views(self => ({
      get name(): string {
        return readConfObject(self.configuration, 'name')
      },
      get aliases(): string[] {
        return readConfObject(self.configuration, 'aliases')
      },
      get refNames() {
        return self.regions.map(region => region.refName)
      },
      get ready() {
        return !!self.regions.length && self.refNameAliasesSet
      },
      get allRefNames() {
        const aliases: string[] = []
        self.refNameAliases.forEach(aliasList => {
          aliases.push(...aliasList)
        })
        return [...this.refNames, ...aliases]
      },
      get rpcManager() {
        return getParent(self, 2).rpcManager
      },
      getCanonicalRefName(refName: string) {
        if (this.refNames.includes(refName)) {
          return refName
        }
        for (const [rName, aliases] of self.refNameAliases) {
          if (aliases.includes(refName)) {
            return rName
          }
        }
        return undefined
      },
      isValidRefName(refName: string) {
        return this.allRefNames.includes(refName)
      },
    }))
    .actions(self => ({
      afterAttach() {
        addDisposer(
          self,
          autorun(async () => {
            const sequenceAdapterConfig = readConfObject(self.configuration, [
              'sequence',
              'adapter',
            ])
            const adapterConfigId = jsonStableStringify(sequenceAdapterConfig)
            const sequenceAborter = new AbortController()
            const { rpcManager } = getParent(self, 2)
            rpcManager
              .call(
                adapterConfigId,
                'getRegions',
                {
                  sessionId: self.name,
                  adapterConfig: sequenceAdapterConfig,
                  signal: sequenceAborter.signal,
                },
                { timeout: 1000000 },
              )
              .then((adapterRegions: Region[]) => {
                const adapterRegionsWithAssembly = adapterRegions.map(
                  adapterRegion => {
                    const { refName } = adapterRegion
                    checkRefName(refName)
                    return {
                      ...adapterRegion,
                      assemblyName: self.name,
                    }
                  },
                )
                this.setRegions(adapterRegionsWithAssembly)
              })
              .catch((error: Error) => {
                getParent(self, 3).setError(String(error))
              })
            if (self.configuration.refNameAliases) {
              const refNameAliasesAdapterConfig = readConfObject(
                self.configuration,
                ['refNameAliases', 'adapter'],
              )
              const refNameAliasesAborter = new AbortController()
              rpcManager
                .call(
                  adapterConfigId,
                  'getRefNameAliases',
                  {
                    sessionId: self.name,
                    adapterConfig: refNameAliasesAdapterConfig,
                    signal: refNameAliasesAborter.signal,
                  },
                  { timeout: 1000000 },
                )
                .then(
                  (
                    refNameAliasesList: {
                      refName: string
                      aliases: string[]
                    }[],
                  ) => {
                    const refNameAliases: Record<string, string[]> = {}
                    refNameAliasesList.forEach(refNameAlias => {
                      refNameAlias.aliases.forEach(alias => {
                        checkRefName(alias)
                      })
                      refNameAliases[refNameAlias.refName] =
                        refNameAlias.aliases
                    })
                    this.setRefNameAliases(refNameAliases)
                    this.setRefNameAliasesSet(true)
                  },
                )
                .catch((error: unknown) => {
                  getParent(self, 3).setError(String(error))
                })
            } else {
              this.setRefNameAliasesSet(true)
            }
          }),
        )
      },
      setRegions(regions: Region[]) {
        self.regions = cast(regions)
      },
      setRefNameAliases(refNameAliases: Record<string, string[]>) {
        self.refNameAliases = cast(refNameAliases)
      },
      setRefNameAliasesSet(isSet: boolean) {
        self.refNameAliasesSet = isSet
      },
    }))
    .actions(self => {
      const refNameMapCache = new AbortablePromiseCache({
        // QuickLRU is a good backing cache to use, but you can use any
        // cache as long as it supports `get`, `set`, `delete`, and `keys`.
        cache: new QuickLRU({ maxSize: 1000 }),

        // the `fill` callback will be called for a cache miss
        async fill(
          { adapterConf }: { adapterConf: unknown },
          signal?: AbortSignal,
        ) {
          const refNameMap = observable.map({})
          if (!self.refNameAliasesSet) {
            return refNameMap
          }
          const stateGroupName = jsonStableStringify(adapterConf)

          const refNames = await self.rpcManager.call(
            stateGroupName,
            'getRefNames',
            {
              sessionId: self.name,
              adapterConfig: adapterConf,
              signal,
            },
            { timeout: 1000000 },
          )
          refNames.forEach((refName: string) => {
            checkRefName(refName)
            const aliases = self.refNameAliases.get(refName)
            if (aliases) {
              aliases.forEach(refNameAlias => {
                refNameMap.set(refNameAlias, refName)
              })
            } else {
              self.refNameAliases.forEach((configAliases, configRefName) => {
                if (configAliases.includes(refName)) {
                  refNameMap.set(configRefName, refName)
                  configAliases.forEach(refNameAlias => {
                    if (refNameAlias !== refName)
                      refNameMap.set(refNameAlias, refName)
                  })
                }
              })
            }
          })
          return refNameMap
        },
      })

      /**
       * get Map of `canonical-name -> adapter-specific-name`
       */
      async function getRefNameMapForAdapter(
        adapterConf: unknown,
        opts: { signal?: AbortSignal } = {},
      ) {
        const adapterConfigId = jsonStableStringify(adapterConf)
        return refNameMapCache.get(
          adapterConfigId,
          { adapterConf },
          opts.signal,
        )
      }

      /**
       * get Map of `adapter-specific-name -> canonical-name`
       */
      async function getReverseRefNameMapForAdapter(
        adapterConf: unknown,
        opts: { signal?: AbortSignal } = {},
      ) {
        const map = await getRefNameMapForAdapter(adapterConf, opts)
        const reversed = new Map()
        for (const [canonicalName, adapterName] of map) {
          reversed.set(adapterName, canonicalName)
        }
        return reversed
      }
      return { getRefNameMapForAdapter, getReverseRefNameMapForAdapter }
    })
}

export default function assemblyManagerFactory(assemblyConfigType: IAnyType) {
  return types
    .model({
      assemblies: types.array(assemblyFactory(assemblyConfigType)),
    })
    .views(self => ({
      get(assemblyName: string) {
        const canonicalName = this.aliasMap.get(assemblyName)
        return self.assemblies.find(
          assembly => assembly.name === (canonicalName || assemblyName),
        )
      },
      get aliasMap() {
        const aliases: Map<string, string> = new Map()
        self.assemblies.forEach(assembly => {
          if (assembly.aliases.length) {
            assembly.aliases.forEach(assemblyAlias => {
              aliases.set(assemblyAlias, assembly.name)
            })
          }
        })
        return aliases
      },
      get rpcManager() {
        return getParent(self).rpcManager
      },
      get allPossibleRefNames() {
        const refNames: string[] = []
        self.assemblies.forEach(assembly => {
          refNames.push(...assembly.refNames)
        })
        return refNames
      },
    }))
    .views(self => ({
      async getRefNameMapForAdapter(
        adapterConf: unknown,
        assemblyName: string,
        opts: { signal?: AbortSignal } = {},
      ) {
        const assembly = self.get(assemblyName)
        if (assembly) {
          return assembly.getRefNameMapForAdapter(adapterConf, opts)
        }
        return observable.map({})
      },
      async getReverseRefNameMapForAdapter(
        adapterConf: unknown,
        assemblyName: string,
        opts: { signal?: AbortSignal } = {},
      ) {
        const assembly = self.get(assemblyName)
        if (assembly) {
          return assembly.getReverseRefNameMapForAdapter(adapterConf, opts)
        }
        return observable.map({})
      },
      isValidRefName(refName: string, assemblyName?: string) {
        if (assemblyName) {
          const assembly = self.get(assemblyName)
          if (assembly) {
            return assembly.isValidRefName(refName)
          }
        }
        return self.allPossibleRefNames.includes(refName)
      },
    }))
    .actions(self => ({
      afterAttach() {
        // self.assemblies.push({
        //   configuration: getParent(self).jbrowse.assemblies[0],
        // })
        addDisposer(
          self,
          autorun(() => {
            getParent(self).jbrowse.assemblies.forEach(
              (assemblyConfig: AnyConfigurationModel) => {
                const existingAssemblyIdx = self.assemblies.findIndex(
                  assembly =>
                    assembly.name === readConfObject(assemblyConfig, 'name'),
                )
                if (existingAssemblyIdx !== -1) {
                  this.replaceAssembly(existingAssemblyIdx, assemblyConfig)
                } else {
                  this.addAssembly(assemblyConfig)
                }
              },
            )
          }),
        )
      },
      addAssembly(
        assemblyConfig: SnapshotOrInstance<typeof assemblyConfigType> | string,
      ) {
        self.assemblies.push({ configuration: assemblyConfig })
      },
      replaceAssembly(
        idx: number,
        assemblyConfig: SnapshotOrInstance<typeof assemblyConfigType> | string,
      ) {
        self.assemblies[idx] = assemblyConfigType.create({
          configuration: assemblyConfig,
        })
      },
    }))
}
