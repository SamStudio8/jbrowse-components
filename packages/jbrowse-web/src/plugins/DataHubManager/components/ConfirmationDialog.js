import { TrackDbFile } from '@gmod/ucsc-hub'
import Button from '@material-ui/core/Button'
import Icon from '@material-ui/core/Icon'
import LinearProgress from '@material-ui/core/LinearProgress'
import Typography from '@material-ui/core/Typography'
import { transaction } from 'mobx'
import { observer } from 'mobx-react'
import { onPatch } from 'mobx-state-tree'
import PropTypes from 'prop-types'
import React from 'react'
import JBrowse from '../../../JBrowse'
import Contents from '../../HierarchicalTrackSelectorDrawerWidget/components/Contents'

const supportedTrackTypes = [
  'bam',
  // 'bed',
  // 'bed5FloatScore',
  // 'bedGraph',
  // 'bedRnaElements',
  // 'bigBarChart',
  // 'bigBed',
  // 'bigChain',
  // 'bigInteract',
  // 'bigMaf',
  // 'bigPsl',
  // 'bigWig',
  // 'broadPeak',
  // 'coloredExon',
  // 'gvf',
  // 'ld2',
  // 'narrowPeak',
  // 'peptideMapping',
  // 'vcfTabix',
  // 'wig',
  // 'wigMaf',
]

function makeTrackConfig(track, categories, ignoreUnsupported) {
  const trackType = track.get('type')
  const baseTrackType = trackType.split(' ')[0]
  switch (baseTrackType) {
    case 'bam':
      return {
        type: 'AlignmentsTrack',
        name: track.get('shortLabel'),
        description: track.get('longLabel'),
        category: categories,
        adapter: {
          type: 'BamAdapter',
          bamLocation: { uri: track.get('bigDataUrl') },
          index: {
            location: {
              uri:
                track.get('bigDataIndex') || `${track.get('bigDataUrl')}.bai`,
            },
          },
        },
      }
    default:
      if (ignoreUnsupported)
        return {
          type: 'AlignmentsTrack',
          name: track.get('shortLabel'),
          description: track.get('longLabel'),
          category: categories,
        }
      throw new Error(`Unsupported track type: ${baseTrackType}`)
  }
}

@observer
class ConfirmationDialog extends React.Component {
  static propTypes = {
    trackDbUrl: PropTypes.instanceOf(URL).isRequired,
    assemblyName: PropTypes.string.isRequired,
    hubName: PropTypes.string.isRequired,
    setPatches: PropTypes.func.isRequired,
    enableNext: PropTypes.func.isRequired,
  }

  state = {
    errorMessage: '',
    trackDb: new Map(),
    model: null,
    unsupportedTrackTypes: new Set(),
    unsupportedTrackTypeModels: new Map(),
    renderUnsupportedTrackTypes: false,
  }

  componentDidMount() {
    this.getTrackDb()
  }

  async getTrackDb() {
    const { trackDbUrl, enableNext } = this.props
    let response
    try {
      response = await fetch(trackDbUrl)
    } catch (error) {
      this.setState({
        errorMessage: (
          <span>
            <strong>Network error.</strong> {error.message} <br />
            {trackDbUrl.href}
          </span>
        ),
      })
      return
    }
    if (!response.ok) {
      this.setState({
        errorMessage: (
          <span>
            <strong>Could not access TrackDb file</strong> <br />
            {trackDbUrl.href} <br />
            {response.status}: {response.statusText}
          </span>
        ),
      })
      return
    }
    const responseText = await response.text()
    let trackDb
    try {
      trackDb = new TrackDbFile(responseText)
    } catch (error) {
      console.log(error)
      this.setState({
        errorMessage: (
          <span>
            <strong>Could not parse trackDb.txt file</strong> <br />
            {error.message} <br />
            {trackDbUrl.href}
          </span>
        ),
      })
      return
    }
    const model = this.generateModel(trackDb)
    this.setState({ model, trackDb })
    enableNext()
  }

  toggleUnsupported = () => {
    const {
      renderUnsupportedTrackTypes,
      unsupportedTrackTypeModels,
    } = this.state
    if (!renderUnsupportedTrackTypes && unsupportedTrackTypeModels.size === 0)
      this.generateUnsupportedTrackTypeModels()
    this.setState({ renderUnsupportedTrackTypes: !renderUnsupportedTrackTypes })
  }

  generateUnsupportedTrackTypeModels() {
    const {
      unsupportedTrackTypes,
      trackDb,
      unsupportedTrackTypeModels,
    } = this.state
    unsupportedTrackTypes.forEach(trackType => {
      const model = this.generateModel(trackDb, trackType)
      unsupportedTrackTypeModels.set(trackType, model)
    })
    this.setState({ unsupportedTrackTypeModels })
  }

  generateModel(trackDb, trackType) {
    const { unsupportedTrackTypes } = this.state
    const { assemblyName, hubName, setPatches } = this.props
    const categoryName = `${hubName}: ${assemblyName}`

    const tracks = []

    trackDb.forEach((track, trackName) => {
      const trackKeys = Array.from(track.keys())
      const parentTrackKeys = [
        'superTrack',
        'compositeTrack',
        'container',
        'view',
      ]
      if (trackKeys.some(key => parentTrackKeys.includes(key))) return
      const ucscTrackType = track.get('type').split(' ')[0]
      if (trackType && trackType !== ucscTrackType) return
      if (!trackType && !supportedTrackTypes.includes(ucscTrackType)) {
        unsupportedTrackTypes.add(ucscTrackType)
        return
      }
      const parentTracks = []
      let currentTrackName = trackName
      do {
        currentTrackName = trackDb.get(currentTrackName).get('parent')
        if (currentTrackName) {
          ;[currentTrackName] = currentTrackName.split(' ')
          parentTracks.push(trackDb.get(currentTrackName).get('shortLabel'))
        }
      } while (currentTrackName)
      parentTracks.reverse()
      const categories = [categoryName].concat(parentTracks)
      tracks.push(makeTrackConfig(track, categories, !!trackType))
    })

    const jbrowse = new JBrowse().configure()

    const { model: rootModel } = jbrowse
    const patches = []
    onPatch(rootModel.configuration.tracks, patch => patches.push(patch))

    transaction(() => {
      tracks.forEach(track =>
        rootModel.configuration.addTrackConf(track.type, track),
      )
    })

    setPatches(patches)
    this.setState({ unsupportedTrackTypes })
    const firstView = rootModel.addView('LinearGenomeView')
    firstView.activateTrackSelector()
    return rootModel.drawerWidgets.get('hierarchicalTrackSelector')
  }

  async doGet(url) {
    let rawResponse
    try {
      rawResponse = await fetch(url)
    } catch (error) {
      this.setState({
        errorMessage: (
          <span>
            <strong>Network error.</strong> {error.message} <br />
            {url.href}
          </span>
        ),
      })
      return null
    }
    if (!rawResponse.ok) {
      this.setState({
        errorMessage: (
          <span>
            <strong>URL is invalid</strong> <br />
            {url.href}
          </span>
        ),
      })
      return null
    }
    return rawResponse.text()
  }

  render() {
    const {
      errorMessage,
      model,
      unsupportedTrackTypes,
      unsupportedTrackTypeModels,
      renderUnsupportedTrackTypes,
    } = this.state
    if (errorMessage)
      return <Typography color="error">{errorMessage}</Typography>
    if (!model) return <LinearProgress variant="query" />
    const confirmationContents = [
      <Contents key="mainContent" model={model} category={model.hierarchy} />,
    ]
    if (unsupportedTrackTypes.size) {
      confirmationContents.push(
        <div key="unsupportedMessage">
          <br />
          <Icon style={{ color: 'red' }}>warning</Icon>
          <Typography>
            Some track types in this hub are not currently supported by JBrowse
            and cannot be imported.
          </Typography>
          <Button variant="outlined" onClick={this.toggleUnsupported}>
            {renderUnsupportedTrackTypes ? 'Hide' : 'Show'} unsupported tracks
          </Button>
        </div>,
      )
      if (renderUnsupportedTrackTypes)
        confirmationContents.push(
          <div key="unsupportedList">
            {Array.from(unsupportedTrackTypes.values()).map(trackType => (
              <div key={trackType}>
                <Typography variant="h6">{trackType}</Typography>
                {unsupportedTrackTypeModels.get(trackType) ? (
                  <Contents
                    key="mainContent"
                    model={unsupportedTrackTypeModels.get(trackType)}
                    category={
                      unsupportedTrackTypeModels.get(trackType).hierarchy
                    }
                  />
                ) : null}
              </div>
            ))}
          </div>,
        )
    }
    return confirmationContents
  }
}

export default ConfirmationDialog
