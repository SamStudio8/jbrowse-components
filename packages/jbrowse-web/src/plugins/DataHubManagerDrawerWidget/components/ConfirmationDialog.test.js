import { createMount, createShallow } from '@material-ui/core/test-utils'
import { clone } from 'mobx-state-tree'
import React from 'react'
import JBrowse from '../../../JBrowse'
import ConfirmationDialog from './ConfirmationDialog'

describe('<ConfirmationDialog />', () => {
  let shallow
  let mount
  //   let classes
  let jbrowse
  let rootModel

  beforeAll(() => {
    shallow = createShallow({ untilSelector: 'ConfirmationDialog' })
    mount = createMount()
    // classes = getClasses(<ConfirmationDialog rootModel={rootModel} />)
    jbrowse = new JBrowse().configure({
      configId: 'testing',
    })
    rootModel = jbrowse.model
    const firstView = rootModel.addView('LinearGenomeView', {
      id: 'viewTestingId',
    })
    firstView.activateTrackSelector()
  })

  afterAll(() => {
    mount.cleanUp()
  })

  it('shallowly renders', () => {
    const wrapper = shallow(
      <ConfirmationDialog
        trackDbUrl={new URL('http://localhost:3000/trackDb.txt')}
        assemblyName="hg38"
        hubName="TestHub"
        enableNext={() => {}}
        rootModel={clone(rootModel)}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('mounts', async () => {
    const mockFetch = () =>
      Promise.resolve(
        new Response(
          `track dnaseSignal
bigDataUrl dnaseSignal.bigWig
shortLabel DNAse Signal
longLabel Depth of alignments of DNAse reads
type bigWig

track dnaseReads
bigDataUrl dnaseReads.bam
shortLabel DNAse Reads
longLabel DNAse reads mapped with MAQ
type bam
`,
        ),
      )
    jest.spyOn(global, 'fetch').mockImplementation(mockFetch)
    const wrapper = mount(
      <ConfirmationDialog
        trackDbUrl={new URL('http://test.com/hg19/trackDb.txt')}
        assemblyName="hg38"
        hubName="TestHub"
        enableNext={() => {}}
        rootModel={clone(rootModel)}
      />,
    )
    const instance = wrapper.children().instance().wrappedInstance
    await instance.componentDidMount()
    instance.toggleUnsupported()
    expect(wrapper).toMatchSnapshot()
  })

  it('handles 404', async () => {
    const mockFetch = () => Promise.resolve(new Response('', { status: 404 }))
    jest.spyOn(global, 'fetch').mockImplementation(mockFetch)
    const wrapper = mount(
      <ConfirmationDialog
        trackDbUrl={new URL('http://test.com/trackDb.txt')}
        assemblyName="hg38"
        hubName="TestHub"
        enableNext={() => {}}
        rootModel={rootModel}
      />,
    )
    const instance = wrapper.children().instance().wrappedInstance
    await instance.componentDidMount()
    expect(wrapper).toMatchSnapshot()
  })

  it('handles fetch error', async () => {
    const mockFetch = () => {
      throw new Error()
    }
    jest.spyOn(global, 'fetch').mockImplementation(mockFetch)
    const wrapper = mount(
      <ConfirmationDialog
        trackDbUrl={new URL('http://test.com/trackDb.txt')}
        assemblyName="hg38"
        hubName="TestHub"
        enableNext={() => {}}
        rootModel={rootModel}
      />,
    )
    const instance = wrapper.children().instance().wrappedInstance
    await instance.componentDidMount()
    expect(wrapper).toMatchSnapshot()
  })
})
