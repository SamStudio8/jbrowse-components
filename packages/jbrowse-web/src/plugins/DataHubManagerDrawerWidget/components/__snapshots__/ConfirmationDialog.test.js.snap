// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfirmationDialog /> mounts 1`] = `
<WithStyles(inject-ConfirmationDialog-with-rootModel)
  assemblyName="hg38"
  enableNext={[Function]}
  hubName="TestHub"
  rootModel={
    Object {
      "configuration": Object {
        "_configId": "testing",
        "views": Object {
          "LinearGenomeView": Object {
            "_configId": "LinearGenomeView",
          },
        },
      },
      "drawerWidgets": Object {
        "hierarchicalTrackSelector": Object {
          "collapsed": Object {},
          "filterText": "",
          "id": "hierarchicalTrackSelector",
          "type": "HierarchicalTrackSelectorDrawerWidget",
          "view": "viewTestingId",
        },
      },
      "drawerWidth": 256,
      "menuBars": Array [],
      "selectedDrawerWidget": "hierarchicalTrackSelector",
      "views": Array [
        Object {
          "bpPerPx": 1,
          "configuration": Object {
            "_configId": "LinearGenomeView",
          },
          "controlsWidth": 100,
          "displayedRegions": Array [],
          "flipped": false,
          "id": "viewTestingId",
          "offsetPx": 0,
          "tracks": Array [],
          "type": "LinearGenomeView",
        },
      ],
    }
  }
  trackDbUrl={"http://test.com/trackDb.txt"}
>
  <inject-ConfirmationDialog-with-rootModel
    assemblyName="hg38"
    classes={
      Object {
        "unsupportedHeader": "inject-ConfirmationDialog-with-rootModel-unsupportedHeader-1",
      }
    }
    enableNext={[Function]}
    hubName="TestHub"
    rootModel={
      Object {
        "configuration": Object {
          "_configId": "testing",
          "views": Object {
            "LinearGenomeView": Object {
              "_configId": "LinearGenomeView",
            },
          },
        },
        "drawerWidgets": Object {
          "hierarchicalTrackSelector": Object {
            "collapsed": Object {},
            "filterText": "",
            "id": "hierarchicalTrackSelector",
            "type": "HierarchicalTrackSelectorDrawerWidget",
            "view": "viewTestingId",
          },
        },
        "drawerWidth": 256,
        "menuBars": Array [],
        "selectedDrawerWidget": "hierarchicalTrackSelector",
        "views": Array [
          Object {
            "bpPerPx": 1,
            "configuration": Object {
              "_configId": "LinearGenomeView",
            },
            "controlsWidth": 100,
            "displayedRegions": Array [],
            "flipped": false,
            "id": "viewTestingId",
            "offsetPx": 0,
            "tracks": Array [],
            "type": "LinearGenomeView",
          },
        ],
      }
    }
    trackDbUrl={"http://test.com/trackDb.txt"}
  >
    <ConfirmationDialog
      assemblyName="hg38"
      classes={
        Object {
          "unsupportedHeader": "inject-ConfirmationDialog-with-rootModel-unsupportedHeader-1",
        }
      }
      enableNext={[Function]}
      hubName="TestHub"
      rootModel={
        Object {
          "configuration": Object {
            "_configId": "testing",
            "views": Object {
              "LinearGenomeView": Object {
                "_configId": "LinearGenomeView",
              },
            },
          },
          "drawerWidgets": Object {
            "hierarchicalTrackSelector": Object {
              "collapsed": Object {},
              "filterText": "",
              "id": "hierarchicalTrackSelector",
              "type": "HierarchicalTrackSelectorDrawerWidget",
              "view": "viewTestingId",
            },
          },
          "drawerWidth": 256,
          "menuBars": Array [],
          "selectedDrawerWidget": "hierarchicalTrackSelector",
          "views": Array [
            Object {
              "bpPerPx": 1,
              "configuration": Object {
                "_configId": "LinearGenomeView",
              },
              "controlsWidth": 100,
              "displayedRegions": Array [],
              "flipped": false,
              "id": "viewTestingId",
              "offsetPx": 0,
              "tracks": Array [],
              "type": "LinearGenomeView",
            },
          ],
        }
      }
      trackDbUrl={"http://test.com/trackDb.txt"}
    >
      <WithStyles(LinearProgress)
        variant="query"
      >
        <LinearProgress
          classes={
            Object {
              "bar": "MuiLinearProgress-bar-12",
              "bar1Buffer": "MuiLinearProgress-bar1Buffer-17",
              "bar1Determinate": "MuiLinearProgress-bar1Determinate-16",
              "bar1Indeterminate": "MuiLinearProgress-bar1Indeterminate-15",
              "bar2Buffer": "MuiLinearProgress-bar2Buffer-19",
              "bar2Indeterminate": "MuiLinearProgress-bar2Indeterminate-18",
              "barColorPrimary": "MuiLinearProgress-barColorPrimary-13",
              "barColorSecondary": "MuiLinearProgress-barColorSecondary-14",
              "buffer": "MuiLinearProgress-buffer-7",
              "colorPrimary": "MuiLinearProgress-colorPrimary-3",
              "colorSecondary": "MuiLinearProgress-colorSecondary-4",
              "dashed": "MuiLinearProgress-dashed-9",
              "dashedColorPrimary": "MuiLinearProgress-dashedColorPrimary-10",
              "dashedColorSecondary": "MuiLinearProgress-dashedColorSecondary-11",
              "determinate": "MuiLinearProgress-determinate-5",
              "indeterminate": "MuiLinearProgress-indeterminate-6",
              "query": "MuiLinearProgress-query-8",
              "root": "MuiLinearProgress-root-2",
            }
          }
          color="primary"
          variant="query"
        >
          <div
            className="MuiLinearProgress-root-2 MuiLinearProgress-colorPrimary-3 MuiLinearProgress-query-8"
            role="progressbar"
          >
            <div
              className="MuiLinearProgress-bar-12 MuiLinearProgress-barColorPrimary-13 MuiLinearProgress-bar1Indeterminate-15"
              style={Object {}}
            />
            <div
              className="MuiLinearProgress-bar-12 MuiLinearProgress-barColorPrimary-13 MuiLinearProgress-bar2Indeterminate-18"
              style={Object {}}
            />
          </div>
        </LinearProgress>
      </WithStyles(LinearProgress)>
    </ConfirmationDialog>
  </inject-ConfirmationDialog-with-rootModel>
</WithStyles(inject-ConfirmationDialog-with-rootModel)>
`;

exports[`<ConfirmationDialog /> shallowly renders 1`] = `
<WithStyles(LinearProgress)
  variant="query"
/>
`;
